### ISSUE TITLE
Learning independent length scales for n-D input in Approximate GP

### ISSUE THREAD

**Issue by martinitus** [cornellius-gp/gpytorch#1359]
Hi Folks!

After looking around in the documentation and trying various different combinations of Kernel configurations, I simply have to ask :heart: 

Setup: 
- 3D training data with significantly different length scales in the individual dimensions
- Approximate GP a la
   ```python
    distribution = CholeskyVariationalDistribution(inducing.shape[0])
    strategy = VariationalStrategy(self, inducing, distribution, learn_inducing_locations=True)
    ```
- Kernel configuration
    ```python
    self.mean_module = gpytorch.means.ConstantMean()
    # also tried without the wrapping ScaleKernel and passing as well as various different length scale priors.
    self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(ard_num_dims=3))
    ```

In my experiments the lengthscale vector always changed simultaneousliy in all 3 dimensions. Also in some of the experiments 
it converged towards `ln(2)=0.69314718` which seemed very suspicious to me.

I saw many tutorials where the length scale of the nested base kernel converges during training iteration. However I think none of those examples where using the approximate GP setup. 

 - Can anyone point me to an approximate GP setup with independent lengthscales? 
 - Am I completely wrong and this is simply not supported in approximate GP? Do I have to define them in advance?
 - What (probably obvious) other thing could I do wrong?

Any hints are welcome :+1: - If this is in fact supposed to work, I will gladly provide a minimal example with my "non-working" setup...

Kind regards and keep up your awesome project!
Martin

_BTW: I did not find any mailing list or slack link where I could have asked this. If there is anything like this, maybe it would be worth putting it to the readme :-)_

**Comment by jacobrgardner** [comment]
@martinitus You're somehow just not getting gradients. That lengthscale is the default initialization in gpytorch, e.g. `inv_softplus(1)`.

Can you provide a more complete code example? The two snippets you have posted should (and do, after testing) work just fine.

**Comment by martinitus** [comment]
Maybe it is because I use `BernoulliLikelihood`, `Adam` optimizer, or `VariationalELBO` mll definition.

@jacobrgardner I put together an example of my setup with fake input data below - The printed length scales always are
`[0.6931472 0.6931472]` which is ln(2).

```python
import numpy
import torch
import gpytorch

import gpytorch.kernels

from torch import Tensor
from gpytorch.distributions import MultivariateNormal
from gpytorch.models import ApproximateGP
from gpytorch.variational import CholeskyVariationalDistribution, VariationalStrategy

from typing import List, Union, Optional, Tuple
from torch.utils.data import DataLoader, TensorDataset
from gpytorch.likelihoods import BernoulliLikelihood # <---- maybe it is because of this?!?

import matplotlib.pyplot as plt

print(torch.__version__,gpytorch.__version__) # => torch: 1.7.0, gpytorch: 1.2.1

class GPModel(ApproximateGP):
    """
    Pretty much a copy from the variational GP tutorial.
    """
    def __init__(self, inducing: Tensor): 
        distribution = CholeskyVariationalDistribution(inducing.shape[0])
        strategy = VariationalStrategy(self, inducing, distribution, learn_inducing_locations=True)
        super(GPModel, self).__init__(strategy)
        self.mean_module = gpytorch.means.ConstantMean()
        
        # lengthscale always results in [0.6931472 0.6931472] for both versions
        self.covar_module = gpytorch.kernels.RBFKernel(ard_num_dims=2)
        # self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(ard_num_dims=2))
        
    def forward(self, x: torch.Tensor) -> MultivariateNormal:
        mean = self.mean_module(x)
        covar = self.covar_module(x)
        return MultivariateNormal(mean, covar)

def toy_model_data(samples: int)->Tuple[numpy.ndarray, numpy.ndarray]:
    """
    Return tuple of training x, y data where x is of shape (3, samples) 
    the structure of the underlying function has different length scales
    in different dimensions.
    """
    numpy.random.seed(0)
    torch.random.manual_seed(0)

    def mean(x, y, z):
        """The target function that should be approximated by the GP"""
        m = numpy.cos(x - 3) * ((y + 30) / 10) ** 1.5 + z
        return m / m.max()

    a = numpy.random.normal(loc=3, scale=.5, size=(samples,))
    b = numpy.random.randint(-30, -10, size=(samples,))
    c = numpy.random.random(size=(samples,))

    # choose binary variable (true/false) with probability given by a normal distribution defined by x,y,z
    t = numpy.random.binomial(1, mean(a, b, c))

    return (numpy.column_stack([a, b, c]), t), mean

likelihood = gpytorch.likelihoods.BernoulliLikelihood()
model = GPModel(inducing=torch.Tensor(numpy.random.normal(loc=0, scale=1, size=(30, 2))))

# generate and plot sample data as well as toy function
(x, y), mean = toy_model_data(samples=1000)

amin, amax = 0, 6
bmin, bmax = -30, -10
vmin, vmax = 0, 0.5
N = 50

plt.figure()
m = mean(numpy.linspace(amin, amax, N).reshape(N, 1), numpy.linspace(bmin, bmax, N).reshape(1, N), 0.5)
plt.title("toy function & training data")
plt.imshow(m.T, extent=(amin, amax, bmin, bmax), origin="lower", aspect="auto", vmin=vmin, vmax=vmax,interpolation="none")
plt.scatter(x[:,0],x[:,1],c='k',marker='x',s=3)
plt.xlabel("a")
plt.ylabel("b")
plt.xlim(amin, amax)
plt.ylim(bmin, bmax)
plt.show()

lr = 0.01
n_epochs = 200

loader = DataLoader(
    batch_size= 500,
    shuffle=True,
    dataset=TensorDataset(torch.tensor(x[:,0:2], dtype=torch.float), torch.tensor(y, dtype=torch.float))
   )

model.train()
likelihood.train()
optimizer = torch.optim.Adam(model.parameters(), lr=lr)
mll = gpytorch.mlls.VariationalELBO(likelihood, model, len(x))

losses = []
for i in range(n_epochs):
    losses = []
    for x_batch, y_batch in loader:
        # Zero backpropped gradients from previous iteration
        optimizer.zero_grad()
        # Get predictive (latent) output
        output = model(x_batch)
        # Calculate loss and update
        loss = -mll(output, y_batch)
        losses.append(loss.item())
        loss.backward()
        optimizer.step()
    
    kernel = model.covar_module if isinstance(model.covar_module, gpytorch.kernels.RBFKernel )else model.covar_module.base_kernel

    print(f'Iter {i}: Loss: {loss.item():.3f} lengthscale: {kernel.lengthscale.detach().cpu().numpy()}')
```




I also attached the respective notebook.

[gpytorch length scale learning.ipynb.txt](https://github.com/cornellius-gp/gpytorch/files/5588981/gpytorch.length.scale.learning.ipynb.txt)

**Comment by jacobrgardner** [comment]
@martinitus Normalize your inputs, e.g. z-score `x`, and that'll fix everything. The problem is right now the kernel is getting vanishingly small gradients because our default initialization scheme assumes normalized inputs and is very far off from optimal for the specific scale of data you have there. So you could also alternatively change the lengthscale initialization, but normalizing your inputs would be better, e.g., `x = (x - x.mean(-2)) / x.std(-2)`

One other way I often use to check for / fix this is to just add a BatchNorm feature extractor:
```diff
class GPModel(ApproximateGP):
    """
    Pretty much a copy from the variational GP tutorial.
    """
    def __init__(self, inducing: Tensor): 
        distribution = CholeskyVariationalDistribution(inducing.shape[0])
        strategy = VariationalStrategy(self, inducing, distribution, learn_inducing_locations=True)
        super(GPModel, self).__init__(strategy)
        self.mean_module = gpytorch.means.ConstantMean()
        
        # lengthscale always results in [0.6931472 0.6931472] for both versions
        self.covar_module = gpytorch.kernels.RBFKernel(ard_num_dims=2)
        # self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(ard_num_dims=2))
        
+      self.normalizer = torch.nn.BatchNorm1d(inducing.size(-1))
        
    def forward(self, x: torch.Tensor) -> MultivariateNormal:
+        x = self.normalizer(x)
        mean = self.mean_module(x)
        covar = self.covar_module(x)
        return MultivariateNormal(mean, covar)
```

**Comment by martinitus** [comment]
@jacobrgardner Thank you so much! I actually removed normalization code which was based on sklearn to get rid of the sklearn dependency. Now at least I know how to achieve something similar with torch 👎🏼 )
